### ISSUE TITLE
Tar download corrupt with big files

### ISSUE THREAD

**Issue by mivk** [filesender/filesender#2062]
When a big file is downloaded as .tar archive, the resulting "transfer_... .tar" file is corrupt. When downloaded as .zip it is ok. And with smaller files, the .tar is also ok. In my case, the file triggering the problem is 71 GB. I haven't yet done more tests to see what the size limit is.

The `tar --test` command shows the files that should be in the archive, but with a wrong size for the big file, and then come tons of errors and the file cannot be extracted.

The files I selected in the download window are:

* ASSETMAP.xml 1.8 kB
* j2c_d89a0264-f3eb-44bf-869e-896e9aa00c7c.mxf 71.1 GB

The result of `tar -tvf transfer_2335_files_cd4d8310.tar` (note also the wrong size for the "j2c...mxf" file) :

```
-rwxrwxrwx 0/0            1832 2024-10-31 12:01 ASSETMAP.xml
-rwxrwxrwx 0/0      9540470703 2024-10-31 12:01 j2c_d89a0264-f3eb-44bf-869e-896e9aa00c7c.mxf
tar: Skipping to next header
tar: Archive contains `\354`L\234\311S\016\027\346%E1' where numeric off_t value expected
tar: Archive contains `1\231\220\232\f._' where numeric mode_t value expected
tar: Archive contains `]\252?̺\353\303\314\322\035)H' where numeric time_t value expected
tar: Archive contains `\ri˥)q\330' where numeric uid_t value expected
tar: Archive contains `W\350ff\342%\225\242' where numeric gid_t value expected
tar: Archive contains `\ri˥)q\330' where numeric uid_t value expected
tar: Archive contains `W\350ff\342%\225\242' where numeric gid_t value expected
?rwsrwsrwt 4294967295/4294967295 18446744073709551615 1970-01-01 00:59 \322hf[\f\343\374\304\350\220b8\232\353n\202\237\261\264\361\226m-C\2642\026\273s\312\026I(\202nC\201d\255\026\336X\335}k,N\373\333\343\231\355\252n\3243\322\345@\330b\350J unknown file type `\336'
tar: Skipping to next header
tar: Archive contains `a\027\316\215\313\022\310\004\026]J2' where numeric off_t value expected
tar: Archive contains `w\214\332g\252\343\377a' where numeric mode_t value expected
tar: Archive contains `(gu \316x\221\266\360{\336\332' where numeric time_t value expected
tar: Archive contains `\2440\031R\255\377?\205' where numeric uid_t value expected
tar: Archive contains `\316\321}\303t-\230!' where numeric gid_t value expected
tar: Archive contains `\2440\031R\255\377?\205' where numeric uid_t value expected
tar: Archive contains `\316\321}\303t-\230!' where numeric gid_t value expected
?rwsrwsrwt 4294967295/4294967295 18446744073709551615 1970-01-01 00:59 \0027\235\001 unknown file type `\225'
tar: Skipping to next header
```

This repeats 41 times with different content shown, until ending with

```
tar: Skipping to next header
tar: Exiting with failure status due to previous errors
```
This is with filesender version 2.17. But if the functions used are the ones from [lib/vendor/barracudanetworks/archivestream-php/src/TarArchive.php](https://github.com/filesender/filesender/blob/development/lib/vendor/barracudanetworks/archivestream-php/src/TarArchive.php), they seem to not have changed since then.

**Comment by monkeyiq** [comment]
I am guessing the first port of call would be to use the library directly to make a file of that size.

**Comment by mivk** [comment]
Yes. Unfortunately, while I'm familiar with Perl and Bash, I don't really know how to write a little PHP script or command-line to test that.

**Comment by monkeyiq** [comment]
No worries. I can take a look at that.

**Comment by monkeyiq** [comment]
I made a very basic script at 
https://github.com/monkeyiq/filesender/blob/2024/nov/tartest/scripts/testing/tar-test.php

It seems to make small archives ok. I haven't ramped it up to 70gb to see at what point it fails yet.

**Comment by mivk** [comment]
Thank you. Using your script and the many big files I have around, I found that the problem occurs when a single file is over around 70 GB.
The resulting tar file doesn't seem to be limited. Only the size of individual files.

This tar file of over 84 GB is OK:

```
# tar -tvf tar/test4.tar
-rwxrwxrwx 0/0      1672455257 2024-11-02 12:52 RDS_PictLock_20230824-DNx-51-test_5mn.mov
-rwxrwxrwx 0/0     28797780117 2024-11-02 12:52 RDS_PictLock_20230824-DNx-51.mov
-rwxrwxrwx 0/0     53756468750 2024-11-02 12:54 SYSLN_DC_001_100_220121_13_ProRes.mov
-rwxrwxrwx 0/0            1712 2024-11-02 12:58 info.txt
-rwxrwxrwx 0/0        26452075 2024-11-02 12:58 sample.mkv
```

And this one with a 68.01 GB (63.34 GiB) file is also OK:

```
# tar -tvf tar/test6.tar
-rwxrwxrwx 0/0     68012909786 2024-11-02 13:54 814ab685-42fa-46ca-bb4b-843e6489623f_video.mxf
-rwxrwxrwx 0/0            1712 2024-11-02 13:59 info.txt
-rwxrwxrwx 0/0        26452075 2024-11-02 13:59 sample.mkv
```

But this one with a file of 71.46 GB (66.55 GiB) is broken:

```
# ls -l tar/input
-rw-r--r-- 1 root root 71468054665 Nov  2 14:35 747178af-4c8b-4b24-a9dc-8fc17a256f40_video.mxf
-rw-r--r-- 1 mi   mi          1712 Dec  9  2021 info.txt

# tar -tvf tar/test9.tar
-rwxrwxrwx 0/0      8933506833 2024-11-02 18:09 747178af-4c8b-4b24-a9dc-8fc17a256f40_video.mxf
tar: Skipping to next header
tar: Archive contains ‘\260\222G\314g+\022\255\266\022\251\223’ where numeric off_t value expected
^C
```

Before opening this issue, I had done some Googling, and had come across [this very old answer](http://answers.google.com/answers/threadview/id/25116.html#answer) which looks like it might still be relevant in this case:

> The quick answer is: 2^63-1 for the archive, 68'719'476'735 (8^12-1)
> bytes for each file,

> The size of files you put into a tar archive (not the archive itself)
> is limited to 11 octal digits, the max. size of a single file is thus
> ca. 68 GBytes.

**Comment by monkeyiq** [comment]
I imagine if gnu tar can make a file with single files of size 71468054665 then there is some extension to the format that we can also take advantage of. I am in favour of exploring that rather than limiting the size of files that FileSender can allow in tar files.

**Comment by monkeyiq** [comment]
It seems most tar formats want to support Unlimited file sizes...
https://www.gnu.org/software/tar/manual/html_section/Formats.html

**Comment by monkeyiq** [comment]
I see now in the code where the octal size value is truncated to 12 octets in length.  This gives an expected result for a 60gb size value and an invalid response for a 70gb size. GNU tar can archive a 70gb single file happily, so I expect there is an additional header used for the size there but haven't found it in hexl as yet.

**Comment by monkeyiq** [comment]
It seems that gnu tar stores the size in network byte order binary for large files. I haven't checked if it does this for small ones too. There are 12 bytes reserved for the size and originally the size was stored in octal and then converted to ascii. This limits how large a number can be in 12 bytes. Looking at a gnu tar created for a single 70gb file there are 4 bytes roasted up with an `0x80 0x00 0x00 0x00` in nbo. Then the remaining 8 bytes can be used to store the `long long` and if you are storing a 64bit number in nbo binary you get to store some *really big files*. 

I have pushed an update to the archive generator that switches over to nbo binary for files over 2^63-1 in size. tested this first against a file that was a dd from /dev/zero to 70gb in size. Worked ok. Then created a file with many copies of `Fedora-KDE-Live-x86_64-41-1.4.iso` to get to around 84gb in size. Repeating the below a number of times to get to that point. That should give the md5sum if you have the same number of iterations :)

```
dd if=Fedora-KDE-Live-x86_64-41-1.4.iso of=bigtestfile bs=1G oflag=append conv=notrunc
md5sum  bigtestfile 
a1f28fa449d328ef410e207ad15040b3  bigtestfile
```

The test script made a compatible tar file that gnu tar was happy to view and extract

```
$ mkdir -p testdir
$ mv bigtestfile testdir
$ php /opt/filesender/scripts/testing/tar-test.php made-with-filesender.tar testdir
$ mkdir out
$ cd out
$ tar tvf ../made-with-filesender.tar 
-rwxrwxrwx 0/0     89804242944 2024-11-06 09:22 bigtestfile
$ tar xvf ../made-with-filesender.tar 
bigtestfile
$ md5sum  bigtestfile 
a1f28fa449d328ef410e207ad15040b3  bigtestfile
```

So this appears to work ok. I am of the mind that if gnu tar can extract it ok then that is the pillar that I base "works ok" on.

**Comment by monkeyiq** [comment]
Thank you for reporting this. I might roll the patch from the above branch into the next release.

**Comment by monkeyiq** [comment]
With the above (now merged) PR I created a tar file with an 84gb file in it. It expanded fine with gnu tar and the md5 sum matches on the expanded and original data. A 16kb small file was in there too and that also expanded fine and checked. This will be in the next release.

**Comment by madsi1m** [comment]
Can we get this (and other changes) merged into dev3 as well?

**Comment by monkeyiq** [comment]
I am planning a release in soon (1-2 weeks) with this and some other changes in it.

**Comment by monkeyiq** [comment]
@madsi1m I just did a git merge for dev3. I also updated the deps for it. There is still a change to come pending a new rc.
